Source/WebCore/dom/CustomElementReactionQueue.h

```diff
@@ -40,29 +42,74 @@ class CallFrame;
 
 namespace WebCore {
 
-class CustomElementReactionQueueItem;
+class CustomElementQueue;
 class Document;
 class Element;
 class HTMLFormElement;
 class JSCustomElementInterface;
-class QualifiedName;
+
+class CustomElementReactionQueueItem {
+    WTF_MAKE_FAST_ALLOCATED;
+    WTF_MAKE_NONCOPYABLE(CustomElementReactionQueueItem);
+public:
+    enum class Type : uint8_t {
+        Invalid = 0,
+        ElementUpgrade,
+        Connected,
+        Disconnected,
+        Adopted,
+        AttributeChanged,
+        FormAssociated,
+        FormReset,
+        FormDisabled,
+        FormStateRestore,
+    };
+
+    struct AdoptedPayload {
+        Ref<Document> oldDocument;
+        Ref<Document> newDocument;
+        ~AdoptedPayload();
+    };
+
+    struct FormAssociatedPayload {
+        RefPtr<HTMLFormElement> form;
+        ~FormAssociatedPayload();
+    };
+
+    using AttributeChangedPayload = std::tuple<QualifiedName, AtomString, AtomString>;
+    using FormDisabledPayload = bool;
+    using FormStateRestorePayload = CustomElementFormValue;
+    using Payload = std::optional<std::variant<AdoptedPayload, AttributeChangedPayload, FormAssociatedPayload, FormDisabledPayload, FormStateRestorePayload>>;
+
+    CustomElementReactionQueueItem();
+    CustomElementReactionQueueItem(CustomElementReactionQueueItem&&);
+    CustomElementReactionQueueItem(Type, Payload = std::nullopt);
+    ~CustomElementReactionQueueItem();
+    Type type() const { return m_type; }
+    void invoke(Element&, JSCustomElementInterface&);
+
+private:
+    Type m_type;
```

Comment:
Seems we could use `{ Type::Invalid }` here.

Reply:
Will do.